File: DuDoong-Api/src/main/java/band/gosrock/api/order/service/CreateOrderUseCase.java
package band.gosrock.api.order.service;


import band.gosrock.api.common.UserUtils;
import band.gosrock.api.order.model.dto.request.CreateOrderRequest;
import band.gosrock.api.order.model.dto.response.CreateOrderResponse;
import band.gosrock.api.order.model.mapper.OrderMapper;
import band.gosrock.common.annotation.UseCase;
import band.gosrock.domain.domains.order.service.CreateOrderService;
import band.gosrock.domain.domains.user.domain.User;
import lombok.RequiredArgsConstructor;

@UseCase
@RequiredArgsConstructor
public class CreateOrderUseCase {

    private final CreateOrderService createOrderService;

    private final UserUtils userUtils;

    private final OrderMapper orderMapper;

    public CreateOrderResponse execute(CreateOrderRequest createOrderRequest) {
        User user = userUtils.getCurrentUser();
        Long couponId = createOrderRequest.getCouponId();
        Long cartId = createOrderRequest.getCartId();
        if (couponId == null) {
            return orderMapper.toCreateOrderResponse(
                    createOrderService.withOutCoupon(cartId, user.getId()));
        }
        return orderMapper.toCreateOrderResponse(
                createOrderService.withCoupon(cartId, user.getId(), couponId));
    }
}

File: DuDoong-Api/src/main/kotlin/band/gosrock/api/order/service/CreateOrderUseCase.kt
package band.gosrock.api.order.service

import band.gosrock.api.order.model.dto.request.CreateOrderRequest
import band.gosrock.api.order.model.dto.response.CreateOrderResponse
import band.gosrock.api.order.model.mapper.OrderMapper
import band.gosrock.common.annotation.UseCase
import band.gosrock.domain.domains.order.service.CreateOrderService
import org.slf4j.LoggerFactory

@UseCase
class CreateOrderUseCase(
    private val createOrderService: CreateOrderService,
    private val orderMapper: OrderMapper,
) {
    private val log = LoggerFactory.getLogger(CreateOrderUseCase::class.java)

    fun execute(userId: Long, createOrderRequest: CreateOrderRequest): CreateOrderResponse {
        log.info("[CreateOrderUseCase][execute] 주문 생성 userId={} cartId={} couponId={}", userId, createOrderRequest.cartId, createOrderRequest.couponId)
        val couponId = createOrderRequest.couponId
        val cartId = createOrderRequest.cartId!!
        return if (couponId == null) {
            orderMapper.toCreateOrderResponse(createOrderService.withOutCoupon(cartId, userId))
        } else {
            orderMapper.toCreateOrderResponse(createOrderService.withCoupon(cartId, userId, couponId))
        }
    }
}
